File: mobile/android/app/src/main/java/com/mesh/wallet/data/tron/TronAmountEncoder.kt
package com.mesh.wallet.data.tron

import com.mesh.wallet.core.config.TronConfiguration
import java.math.BigDecimal
import java.math.RoundingMode

object TronAmountEncoder {
    fun usdtToSmallestUnits(amount: BigDecimal): Long {
        require(amount > BigDecimal.ZERO) { "Invalid amount" }
        val scaled = amount.multiply(BigDecimal.TEN.pow(TronConfiguration.TOKEN_DECIMALS))
            .setScale(0, RoundingMode.HALF_UP)
        val value = scaled.longValueExact()
        require(value > 0) { "Invalid amount" }
        return value
    }

    fun smallestUnitsToUsdt(smallestUnits: Long): BigDecimal =
        BigDecimal(smallestUnits).divide(BigDecimal.TEN.pow(TronConfiguration.TOKEN_DECIMALS))

    fun encodeUInt256(smallestUnits: Long): ByteArray {
        val data = ByteArray(32)
        var value = smallestUnits
        var index = 31
        while (value > 0 && index >= 0) {
            data[index] = (value and 0xff).toByte()
            value = value ushr 8
            index--
        }
        return data
    }
}

File: mobile/ios/Mesh/Services/Tron/TronAmountEncoder.swift
import Foundation

enum TronAmountEncoder {
    static func usdtToSmallestUnits(_ amount: Decimal) throws -> UInt64 {
        guard amount > 0 else { throw TronAPIError.invalidAmount }
        let scaled = amount * Decimal(pow(10, Double(TronConfiguration.tokenDecimals)))
        let rounded = (scaled as NSDecimalNumber).rounding(accordingToBehavior: nil)
        let value = rounded.uint64Value
        guard value > 0 else { throw TronAPIError.invalidAmount }
        return value
    }

    static func smallestUnitsToUSDT(_ smallestUnits: UInt64) -> Decimal {
        Decimal(smallestUnits) / Decimal(pow(10, Double(TronConfiguration.tokenDecimals)))
    }

    static func encodeUInt256(smallestUnits: UInt64) -> Data {
        var data = Data(repeating: 0, count: 32)
        var value = smallestUnits
        var index = 31
        while value > 0, index >= 0 {
            data[index] = UInt8(value & 0xff)
            value >>= 8
            index -= 1
        }
        return data
    }
}
